src/interrupt/mod.rs

```diff
@@ -0,0 +1,207 @@
+#[derive(Debug)]
+pub enum Error {
+    /// Operation not supported for this interrupt.
+    OperationNotSupported,
+
+    /// The specified configuration is not valid.
+    InvalidConfiguration,
+
+    /// The interrupt was not enabled.
+    InterruptNotChanged,
+
+    /// The interrupt could not be triggered.
+    InterruptNotTriggered,
+}
+
+impl std::error::Error for Error {}
+
+/// Reuse std::io::Result to simplify interoperability among crates.
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Interrupt error: ")?;
+        match self {
+            Error::OperationNotSupported => write!(f, "operation not supported"),
+            Error::InvalidConfiguration => write!(f, "invalid configuration"),
+            Error::InterruptNotChanged => write!(f, "the interrupt state could not be changed"),
+            Error::InterruptNotTriggered => write!(f, "the interrupt could not be triggered"),
+        }
+    }
+}
+
+/// Trait used by interrupt producers to signal interrupts.
+///
+/// An object having the Interrupt Trait is shared between the VMM and the device
+/// that is using the interrupt.
+/// The `Interrupt` performs two main goals:
+///     * offers a control interface for the interrupt through the `enable()`
+///     and `disable()` methods;
+///     * offers a channel through which notifications can be passed between
+///     the VMM and the device;
+///
+/// When an `Interrupt` is triggered, a notification mechanism that is known by the
+/// Hypervisor will be used to signal the guest. The type of the notification mechanism
+/// is defined by the `NotifierType` associated type.
+/// `Interrupt` allows access to the undelying mechanism used by the Hypervisor through
+/// the `notifier()`. This enables some use cases the device may want to bypass the VMM
+/// completely or when the device crate acts only as a control plane and the actual
+/// emulation is implemented in some other component that understands the underlying
+/// mechanism.
+/// A notable example is VFIO that allows a device to register the irqfd so that
+/// interrupts follow a fast path that doesn't require going through the VMM.
+///
+/// Objects implementing this trait are required to have internal mutability.
+pub trait Interrupt {
+    /// The type of the underlying mechanism used for notifications by this interrupt.
+    type NotifierType;
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self) -> Result<()>;
+
+    /// Returns an interrupt notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest through a different interface other
+    /// than `trigger`.
+    fn notifier(&self) -> Option<Self::NotifierType> {
+        None
+    }
+
+    /// Called back when the CPU acknowledges the interrupt.
```

Comment:
What does this mean? The CPU does not acknowledge interrupts.

Is this called when the CPU actually receives the interrupt or when it is logged into the LAPIC or when the guest does an EOI?

Reply:
Need to update the doc a bit here. It's actually an EOI callback and it would be used for shared level-sensitive interrupts in order to inject the interrupt if the device still requires it (e.g. for INTx interrupts). It's equivalent to the `ack_notifier()` the difference being that the latter would enable the resampled semnatics with external components like vfio.

Reply:
For level-triggered interrupts the `trigger` method seems not to be the right abstraction to me. In this case, the device would need a mechanism to assert/de-assert the interrupt line for correct emulation. Optimizations to avoid assert/deassert messages when the interrupt cannot be injected anyway would be a second step.

In any case, the flow for handling level triggered interrupts should be documented. I wouldn't have seen how this is supposed to work from the code alone.